### ISSUE TITLE
Scale controller with Service Bus trigger to ignore scheduled/deferred messages

### ISSUE THREAD

**Issue by SeanFeldman** [Azure/azure-functions-host#7767]
#### What problem would the feature you're requesting solve? Please describe.

Scale controller takes scheduled and deferred messages into consideration, mistaking those messages for active messages and assuming it needs to scale out to process those.

#### Describe the solution you'd like

Do not take scheduled and deferred messages into account when calculating the necessary scaling.

#### Describe alternatives you've considered

Limiting scale-out to keep the cost under control. This is sub-optimal and only prevents over charging to an extent. 

#### Additional context

Customer reported findings: https://discuss.particular.net/t/scheduled-messages-eg-saga-timeouts-causes-over-scale-on-azure-function-consumption-plan/2751/4?u=seanfeldman

**Comment by v-anvari** [comment]
Hi @SeanFeldman , Can you provide a sample repo for this scenario. Also, can you provide us the invocation id, timestamp, and region of the time you have seen this issue.

**Comment by v-anvari** [comment]
After looking into the discussion in the link you have provided, this looks like a design limitation for now. But, to confirm the same, can you explain us the scenario in detail

**Comment by SeanFeldman** [comment]
@v-anvari please see the issue description with the link to the customer problem described in details. It has the scenario details you're looking for.

**Comment by jenspettersson** [comment]
Hello @v-anvari. As I'm the author of the original post in the link posted by @SeanFeldman, perhaps I could help answering questions. As Sean said, the problem description is described in the link, but this morning I started our debug function app again and collected some logs that might help you.

The scenario was:

1. Send one scheduled message for 30 minutes in the future
2. Send one regular, non-scheduled message
3. Wait...


```
1 server online (e09f620ecddc73363f5b6301cd6bcd1b)
2021-10-27T06:19:04.879 Sending message scheduled in 1800 seconds
2021-10-27T06:19:05.008 Executed 'SendScheduled' (Succeeded, Id=4c6f319b-965c-461a-a4f6-15c5b3f27092, Duration=142ms)
...
IDLE WITH 1 SERVER ONLINE
...
2021-10-27T06:23:09.603 Sending message immediately //sending a "regular" non-scheduled message
2021-10-27T06:23:09.614 Executed 'SendImmediate' (Succeeded, Id=fe782302-e10d-4568-b840-87d17b7da700, Duration=70ms)
....
Starts to spawn 4 new instances:

7222f4a4149bdfc611aae2c6561085658f9ee18fce8f00aabbdc086a21b41aa2
73c0ceb3e87000d40f3f094cdd249ca3c43ebf4ad69458d9c02e6f942c566413 
78111f8e840da4007d668432928a55465112f11201543160938a76cee48743a9
f0911ef74cacadb835b79d32a6d23918de1f2f3d770a285ac935159b12461d00
...
IDLE WITH 5 SERVERS ONLINE
```

![image](https://user-images.githubusercontent.com/125493/139023132-d73c8f19-d827-472d-81e3-18906ea2e8f2.png)
_No activity at all for about 30 minutes..._

```
... 30 mins later...
2021-10-27T06:49:04.907 Message: Scheduled message //handling the scheduled message, no more scheduled messages on the queue at this point
2021-10-27T06:50:51.524 DrainMode mode enabled
...
2021-10-27T06:51:05.426 Call to StopAsync complete, registered listeners are now stopped
...
2021-10-27T06:54:19.839 Job host stopped //begin to stop the additional hosts
...
2021-10-27T06:56:17.704 Job host stopped //all additional hosts are stopped, back to 1 running instance
```

Region: West Europe

Is there anything else you need to know?

//J

**Comment by SeanFeldman** [comment]
Linking to the parent/older issue: https://github.com/Azure/Azure-Functions/issues/715

**Comment by cannehag** [comment]
Hello.

We need to move away from Consumption plans due to requirement of virtual network integration, which Elastic plans can provide for us. We currently have 16 FunctionApps with approx. 5-10 function in each of them and as you can imagine, the cost of this will be quite high compared to consumption plans. So we did the initial test of moving some FunctionApps to one Elastic plan to try and figure out how many instances we need.
But, since we rely quite heavily on scheduled messages in some of our FunctionApps this is a big issue since we need to have the maximum amount of instances set to maybe 10 to handle our bursts. But because of this bug/design flaw, it will scale up and use all these 10 instances within some minutes and never scale down since we at all time have some scheduled message in the servicebus.

**Comment by danielmarbach** [comment]
I'm not very familiar with the scale out controller, so this might be a stupid question for those who know the internals. Yet I'm still curious. Isn't this part of the code responsible for doing the scale out?

https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Listeners/ServiceBusScaleMonitor.cs#L109-L110

and that one looks like it doesn't take the TotalCount which therefore should not contain scheduled messages. Or is there something else involved that triggers the behavior we are seeing here?

**Comment by SeanFeldman** [comment]
@danielmarbach wouldn't `TotalCount` count be all the messages, including those that shouldn't be counted? Active messages are the messages ready to be processed, so that code looks ok.

**Comment by danielmarbach** [comment]
Yeah that was my point. It uses Active and Deadlettered so the scale logic should be correct there since those counters don't take scheduled messages into account. Sorry my formulation above is too cryptic

**Comment by jenspettersson** [comment]
This is the information I got from our ticket:

> •	As confirmed with product group, we don’t support this today, and the main issue is that we are using the Peek API, which means Scheduled Queue Messages count in the calculation for the Enqueued Time.

**Comment by danielmarbach** [comment]
Yes that happens here

https://github.com/Azure/azure-sdk-for-net/blob/685f0ffc053b1137c299fa9115df65402b383e26/sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Listeners/ServiceBusScaleMonitor.cs#L68

but this returns the message which has a property `ScheduledEnqueueTime` which could be checked and the logic twisted based on that. 

https://docs.microsoft.com/en-us/dotnet/api/azure.messaging.servicebus.servicebusreceivedmessage.scheduledenqueuetime?view=azure-dotnet#Azure_Messaging_ServiceBus_ServiceBusReceivedMessage_ScheduledEnqueueTime

Basically skipping the analysis of peek messages and peek the next one until a non-deferred message is found. I would need to take a closer look at the code though

**Comment by danielmarbach** [comment]
How about https://github.com/Azure/azure-sdk-for-net/pull/25127 ?

**Comment by FinVamp1** [comment]
@danielmarbach  thanks for the PR. This code fixes the Scale Monitor logic inherent in the Service Bus extension which will be utilised when Runtime Scale Monitoring is enabled which is also known as Virtual Network Triggers support.  Adding @pragnagopa  as we have an internal work item to track the same work for the Scale Controller.

**Comment by SeanFeldman** [comment]
@v-anvari, could you please remove the `Needs: Author Feedback` label to prevent the Stale bot from attempting to close the issue? Thank you.

**Comment by pragnagopa** [comment]
For Consumption SKU - this issue is being addressed in ScaleController. We will update the issue once the fix is rolled out to production.

**Comment by pragnagopa** [comment]
Closing this as dup of https://github.com/Azure/Azure-Functions/issues/715 - will keep issue updated as the fix rolls out.
